### ISSUE TITLE
Error when translating with py-googletrans 4.0.0rc1 (Google Translate backend)

### ISSUE THREAD

**Issue by nexeleriot** [gi-lom/dialect#133]
When I use Google translate it says "Translation failed. Please check for network issues."
But text is translated

![Снимок экрана от 2021-02-20 09-15-57](https://user-images.githubusercontent.com/70663363/108587132-e3287200-7362-11eb-9f7d-209d75ae7f8c.png)
![Снимок экрана от 2021-02-20 09-15-26](https://user-images.githubusercontent.com/70663363/108587134-eae81680-7362-11eb-9e24-6aabca3c78c6.png)

**Comment by fushinari** [comment]
What package are you using? Flatpak, Fedora or AUR?

**Comment by fushinari** [comment]
![Screenshot_20210220_130616](https://user-images.githubusercontent.com/28823274/108587907-6f935e80-737c-11eb-81f7-ab22a72cad76.png)

`main` works for me just fine.

**Comment by LyesSaadi** [comment]
> Edit: Just in case, gonna mention here, py-googletrans-4.0.0rc1 does not work.

Oh, why ?

That is a problem with the Fedora package. I also had these errors, and I thought it was an issue on my end (like API abuse or smth like that) !

**Comment by nexeleriot** [comment]
> What package are you using? Flatpak, Fedora or AUR?

I built it from source yesterday

**Comment by LyesSaadi** [comment]
I'd like to avoid reverting to 3.1.0a... I'll investigate this and try to add compatibility with 4.0.0rc1. I'll also submit it as a PR if you want that.

**Comment by fushinari** [comment]
> I built it from source yesterday

@nexeleriot Did you build it as a flatpak or is it a traditional meson build? If it's the second one, ensure that your py-googletrans is at 3.1.0a and not any other version. It's the only version that works as expected right now.

**Comment by fushinari** [comment]
> Oh, why ?

@LyesSaadi sorry, I should have let you know about this when I first noticed it...

The issue is that the behavior of the `extra_data` variable that we use for the "Mistakes" and "Pronunciation" features is entirely different in 4.0.0rc1.

For example:
3.1.0a `extra_data` for the string "who dit it?" when translated to French:
```python3
{
    "translation": [
        [
            "qui l'a dit?",
            "who dit it?",
            None,
            None,
            3,
            None,
            None,
            [[]],
            [[["ce80756654b03220a807428aa5cafc6b", "en_fr_2020q2.md"]]],
        ]
    ],
    "all-translations": None,
    "original-language": "en",
    "possible-translations": [
        [
            "who dit it?",
            None,
            [["qui l'a dit?", 0, True, False], ["qui dit-il?", 0, True, False]],
            [[0, 11]],
            "who dit it?",
            0,
            0,
        ]
    ],
    "confidence": 0.7623325,
    "possible-mistakes": [
        "who <b><i>did</i></b> it?",
        "who did it?",
        [1],
        None,
        None,
        False,
    ],
    "language": [["en"], None, [0.7623325], ["en"]],
    "synonyms": None,
    "definitions": None,
    "examples": None,
    "see-also": None,
}
```

The same for 4.0.0rc1:
```python3
{
    "confidence": None,
    "parts": [<googletrans.models.TranslatedPart object>],
    "origin_pronunciation": None,
    "parsed": [
        [
            None,
            [[[None, "who <b><i>did</i></b> it?"], None, 2, [1]]],
            "en",
            [[[0, [[[None, 11]], [True]]]], 11],
        ],
        [
            [
                [
                    None,
                    None,
                    None,
                    None,
                    None,
                    [["qui l'a dit?", ["qui l'a dit?", "qui dit-il?"]]],
                ]
            ],
            "fr",
            1,
            "auto",
            ["who dit it?", "auto", "fr", True],
        ],
        "en",
    ],
}
```

It's almost like it went from a well-written poem to pure gibberish... I mean, the info we need is in there somewhere, it's just horribly organized.

So yeah, technically, you can make it work, you just need to check the version of the library and make the necessary changes in `src/translators/gtrans.py` to make the current output adapt to our existing model. But I feel like that work would be pointless until a stable release of 4.0.0 comes.

If you find time to figure out the exact current behavior and change it accordingly, it should work. I don't know if that time's really worth spending though. :sweat_smile: But if you do, send in a PR and I'll gladly merge! :smile:

**Comment by fushinari** [comment]
Btw, I should note that `extra_data` has always been undocumented, which is why I didn't file an issue.

**Comment by nexeleriot** [comment]
Yes, I think that's it. My version is 4.0.0~rc1

**Comment by fushinari** [comment]
Nice! I mean... not nice... but at least we know exactly what's wrong. So we *can* fix it at some point, even if not now.

**Comment by LyesSaadi** [comment]
> If you find time to figure out the exact current behavior and change it accordingly, it should work.

The problem is Fedora's guidelines says nothing about what would happen in case downgrading is necessary. The issue is that with the breakage of 3.0.0, and the release of 3.1.0a I quickly upgraded to it. When 4.0.0 released, I didn't expect it to break. And at that point, I did test the upgrade with the command line utility of google trans which showed no issue. I did IIRC notice the error message, but with what was happening with 3.0.0, I probably thought it was temporary.

So, I think I have no choice but to try and document this!

> I should have let you know about this when I first noticed it...

It's fine, I also should've tested it further. And the app technically still work... Partially :stuck_out_tongue:! I should've also opened an issue at the time.

I think that if googletrans breaks Dialect in another version, you, myself, or anyone, should open an issue as soon as possible to alert others not to update googletrans and to possibly fix this.

**Comment by fushinari** [comment]
> And the app technically still work... Partially :stuck_out_tongue:!

:laughing: Well, preventing crashes from this kind of stuff has kind of been part of our development process. Especially because py-googletrans has been so unreliable in general (not the devs' fault, just Google being Google). It did lead to the main translation function (`run_translation`) being a relatively dirty mess but it does keep the app running somehow the other if possible. :smile: 

> I think that if googletrans breaks Dialect in another version, you, myself, or anyone, should open an issue as soon as possible to alert others not to update googletrans and to possibly fix this.

Yep, so huge thank you to @nexeleriot for bringing my attention back to this. I'd tried to update it when working on #120 and then, just decided to stick with 3.1.0a until 4 reaches stable.

Btw, @LyesSaadi , don't waste time adapting to 4.0.0rc1. I just noticed. 4.0.0rc1 is unusable in Dialect. Other blockers include: https://github.com/ssut/py-googletrans/issues/260 , https://github.com/ssut/py-googletrans/issues/257 , etc. Let's wait for a stable release and then I'll adapt it.

**Comment by LyesSaadi** [comment]
> Btw, @LyesSaadi , don't waste time adapting to 4.0.0rc1. I just noticed. 4.0.0rc1 is unusable even otherwise in Dialect. Other blockers include: ssut/py-googletrans#260 , ssut/py-googletrans#257 , etc. Let's wait for a stable release and then I'll adapt it. For example, translating "next" to French fails.

I guess you're right... I'm going to force Fedora's repos to get back to the old version of googletrans. But I'm still documenting this :stuck_out_tongue:! I think I'll work mainly on simplifying and documenting how Dialect's Translator class's extra_data work to allow other back-ends to effectively use it, and to simplify the handing of googletrans's extra_data updates.

In fact, seeing the code changes, it seems that the extra_data field change may not even be voluntary.

**Comment by LyesSaadi** [comment]
Maybe this issue should be renamed "Error message when translating with Google Translate 4.0.0rc1"

**Comment by fushinari** [comment]
> In fact, seeing the code changes, it seems that the extra_data field change may not even be voluntary.

Actually, I think we can make it work with 4.0.0rc1. Could you wait for a bit?

**Comment by LyesSaadi** [comment]
> In fact, seeing the code changes, it seems that the extra_data field change may not even be voluntary.

My bad, I was looking at the wrong branch...

We can actually make this work using the translate_legacy function !

**Comment by fushinari** [comment]
So... sorry. But again... it doesn't work still. 

Needs merging of https://github.com/ssut/py-googletrans/pull/275 and then a release from their side.

**Comment by LyesSaadi** [comment]
Not a problem, I can patch googletrans.

**Comment by LyesSaadi** [comment]
Anyway, thank you for making this work :tada:!

**Comment by rafaelmardojai** [comment]
How complex is the **py-googletrans** code? Maybe we could fork it and make it part of the dialect translators module.

**Comment by LyesSaadi** [comment]
> How complex is the py-googletrans code? Maybe we could fork it and make it part of the dialect translators module.
Very complex. ~900 lines of code.

**Comment by fushinari** [comment]
> Not a problem, I can patch googletrans.

Awesome! Then I guess that this means it's a non-issue for most people. Btw, can you also test to see if it actually works? I haven't really tested it, I just tested the `translate_legacy` function and didn't check if the changes in Dialect actually work.

> Anyway, thank you for making this work :tada:!

You're welcome and thank *you* for helping me figure this out!

**Comment by LyesSaadi** [comment]
The main problem is that it's not documented. At all.

**Comment by LyesSaadi** [comment]
Actually, looking a bit more into it, it's not _that_ complex, just weirdly designed and documented. But I guess that's Google's fault, not the maintainers'.

**Comment by fushinari** [comment]
@rafaelmardojai There are multiple reasons we'd want to avoid that:

- They have a wider community who fix many issues that pop up. Being an unofficial tool that breaks so often, we need all the support we can get.
- It's a violation of the Google Translate TOS, so we don't want that code in Dialect if possible.
- I'm not sure we want to take on the burden of maintaining that especially when we know it's so fragile.

I did think about making a Dialect organization and separate repositories for each back-end and all but then I realized we don't need all that, especially because we don't have many translation services that we can safely use without violating their TOS. I do still want to document how adding translators into Dialect works though.

@LyesSaadi Things like the reverse engineered token calculation might be hard for us to maintain in case it ever breaks, since we weren't the ones who reverse engineered it. The new `translate` function doesn't need tokens I think (because it's RPC), but then again, the new function doesn't work properly either.

We can try but for the reasons listed above, it might not be such a great idea. If we do fork it, we should go the route of fixing the RPC-based version (i.e. the currently broken 4.0.0rc1 branch).

**Comment by rafaelmardojai** [comment]
Yes, the Google Translate TOS violation is what worries me more. Because we could keep up with upstream changes, but with the advantage of patching and releasing faster.

**Comment by fushinari** [comment]
But then again... if we just keep it in a separate repository, only that repository will be taken down and not the app's repository. And if that happens, we can just drop Google Translate entirely.

**Comment by fushinari** [comment]
So... how about we make a Dialect organization and move this repo there? And then we could fork py-googletrans into it too.

**Comment by LyesSaadi** [comment]
> Btw, can you also test to see if it actually works?

It works perfectly by backporting the patch to Dialect 1.1.1 :tada: !

```patch
diff -urN dialect-1.1.1/src/window.py dialect-1.1.1-patched/src/window.py
--- dialect-1.1.1/src/window.py	2020-12-07 23:19:09.000000000 +0100
+++ dialect-1.1.1-patched/src/window.py	2021-02-20 16:34:59.394894898 +0100
@@ -453,7 +453,7 @@
         self.translate_btn.set_sensitive(self.src_buffer.get_char_count() != 0)
 
     def switch_auto_lang(self, dest_language, src_text, dest_text):
-        src_language = self.translator.detect(src_text).lang
+        src_language = self.translator.detect_legacy(src_text).lang
         if isinstance(src_language, list):
             src_language = src_language[0]
 
@@ -714,7 +714,7 @@
             dest_language = trans_dict['dest_language']
             if src_language == 'auto' and src_text != '':
                 try:
-                    src_language = self.translator.detect(src_text).lang
+                    src_language = self.translator.detect_legacy(src_text).lang
                     if isinstance(src_language, list):
                         src_language = src_language[0]
                     GLib.idle_add(self.src_lang_selector.set_property,
@@ -729,7 +729,7 @@
                 # THIS IS WHERE THE TRANSLATION HAPPENS. The try is necessary to circumvent a bug of the used API
                 if src_text != '':
                     try:
-                        translation = self.translator.translate(
+                        translation = self.translator.translate_legacy(
                             src_text,
                             src=src_language,
                             dest=dest_language
```

**Comment by fushinari** [comment]
Awesome! Glad to know it works.

**Comment by fushinari** [comment]
I'm going to close this since we've fixed the issue from our side as much as possible and the "fix" from our side is out with 1.2.0 (i.e., patching Dialect is not required anymore even though patching py-googletrans still is). We've also sent in a PR with the required patch to upstream. So this is all we can do. :)
